### PR TITLE
Use cookies to perform operations on behalf of the user

### PR DESCRIPTION
**Pull request by AlexandraImbrisca**
- In order to set up the user entities, I associated a separate cookie for sending the id_token to the server.
- I implemented a separate class for getting specific data associated with the user account (e.g. client id, payload).
- The servlets and the DatastoreAccess.java file have been extended to include more functionality (e.g. getting only the groups joined by the user, joining a new group).

Testing and more error checking will be added afterwards.

### CODE REVIEW

File: src/main/java/com/google/lecturechat/data/DatastoreAccess.java [hunk 1]
```diff
@@ -202,4 +236,133 @@ public void addEventToGroup(
...
+
+  /**
+   * Adds the user to the database if they don't exist already.
+   *
+   * @param userId The id of the user that will be added.
+   * @param name The name of the user that will be added.
+   */
+  public void addUser(String userId, String name) {
+    boolean isRegistered = isUserRegistered(userId);
+    if (!isRegistered) {
```
Comment (csperl) [inline src/main/java/com/google/lecturechat/data/DatastoreAccess.java:248]: Avoid nesting by finishing early:

if (!isUserRegistered(userId)) {
  return;
}

File: src/main/java/com/google/lecturechat/data/DatastoreAccess.java [hunk 2]
```diff
@@ -202,4 +236,133 @@ public void addEventToGroup(
...
+      Transaction transaction = datastore.beginTransaction();
```
Comment (csperl) [inline src/main/java/com/google/lecturechat/data/DatastoreAccess.java:249]: I don't think that a transaction is necessary here. Remember in your early classes you simply added entries to the database by calling the put function on the DatastoreService: https://github.com/AlexandraImbrisca/STEP/blob/master/portfolio/src/main/java/com/google/sps/servlets/NewCommentServlet.java#L44

Feel free to read about Transactions and when you need them. https://cloud.google.com/datastore/docs/concepts/transactions

**Review by csperl — CHANGES_REQUESTED** [review]
I did not fully review it yet, but I want to give early feedback about things which apply to most of the code.

Please address them and let me know once you are done :)

File: src/main/java/com/google/lecturechat/data/AuthStatus.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+          String encodedIdToken = cookie.getValue();
+          optionalEncodedIdToken = Optional.of(encodedIdToken);
+        }
+      }
+    }
+
+    if (optionalEncodedIdToken.isPresent()) {
+      String encodedIdToken = optionalEncodedIdToken.get();
+      try {
+        String idToken = java.net.URLDecoder.decode(encodedIdToken, StandardCharsets.UTF_8.name());
```
[the remark this answers is not in the document]
Comment (AlexandraImbrisca) [reply src/main/java/com/google/lecturechat/data/AuthStatus.java:59]: I actually tried without decoding and apparently the function doesn't decode it automatically.

File: src/main/java/com/google/lecturechat/data/AuthStatus.java [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+  /**
+   * Checks if the user is signed by inspecting the cookies included in the request.
+   *
+   * @param request The request from which we will extract the cookies.
+   * @return True if the user is signed in, false otherwise.
+   */
+  public static boolean isSignedIn(HttpServletRequest request) {
+    Optional<String> optionalIdToken = AuthStatus.getIdToken(request);
+
+    if (optionalIdToken.isPresent()) {
```
[the remark this answers is not in the document]
Comment (AlexandraImbrisca) [reply src/main/java/com/google/lecturechat/data/AuthStatus.java:132]: My bad! The id should also be verified :)

File: src/main/java/com/google/lecturechat/data/DatastoreAccess.java (hunk 2 above)
Comment (AlexandraImbrisca) [reply src/main/java/com/google/lecturechat/data/DatastoreAccess.java:249]: I read the documentation again and understood it better. Thank you! :)

**Review by csperl — CHANGES_REQUESTED** [review]
Please add a TODO somewhere that you also verify that the domain of the user who signes-in is a google.com account.

**Comment by AlexandraImbrisca** [comment]
About the user domain, we don't actually have to verify it in the code since we already included this restriction in the OAuth 2.0 consent screen. The user is able to use only Google accounts that have the domain google.com. If the domain is not google.com than the sign in is incomplete and the request is not sent to the server. Do you think we should also include it on the server side?

**Comment by csperl** [comment]
> About the user domain, we don't actually have to verify it in the code since we already included this restriction in the OAuth 2.0 consent screen. The user is able to use only Google accounts that have the domain google.com. If the domain is not google.com than the sign in is incomplete and the request is not sent to the server. Do you think we should also include it on the server side?

I see, sounds good for now.

**Review by csperl — APPROVED** [review]

**Pull request merged**